R/drive_list.R

```diff
@@ -0,0 +1,162 @@
+#' List files on Google Drive
+#'
+#' @param path character vector, path name(s) where the Google drive files are
+#'   that you would like to list. Defaults to the "My Drive" directory.
+#' @param pattern character vector, regular expression(s) of title(s) of
+#'   documents to output in a tibble. If it is `NULL` (default), information
+#'   about all documents in drive will be output in a tibble.
+#' @param ... name-value pairs to query the API
+#' @param verbose logical, indicating whether to print informative messages
+#'   (default `TRUE`)
+#'
+#'   This will default to the most recent 100 files on your Google Drive. If you
+#'   would like more than 100, include the `pageSize` parameter. For example, if
+#'   I wanted 200, I would run `drive_ls(pageSize = 200)`.
+
+#' Helpful links for forming queries:
+#'   * <https://developers.google.com/drive/v3/web/search-parameters>
+#'   * <https://developers.google.com/drive/v3/reference/files/list>
+
+#' @return tibble containing the name, type, and id of files on your google
+#'   drive (default 100 files)
+#' @examples
+#' \dontrun{
+#' ## list user's Google Sheets
+#' drive_list(q = "mimeType='application/vnd.google-apps.spreadsheet'")
+#' }
+#'
+#' @export
+drive_list <- function(path = NULL, pattern = NULL, ..., verbose = TRUE){
+
+  folder <- NULL
+  if (!is.null(path)){
+    folder_pattern <- paste0("^", gsub("/", "$|^", path), "$")
+    folder_nms <- unlist(strsplit(path, "/"))
+    folder_order <- tibble::tibble(name = folder_nms,
+                                   dir = 1:length(folder_nms))
+    folder_tbl <- drive_list(pattern = folder_pattern,
+                             fields = paste0("files/parents,files/name,files/mimeType,files/id"),
+                             q = "mimeType='application/vnd.google-apps.folder'")
+
+    if (!all(folder_nms %in% folder_tbl$name)){
+      spf("We could not find the file path '%s' on your Google Drive", path)
```

Comment:
Technically, this should probably glue the path components up to and including the first one that is not found. So if we're looking for `foo/bar/baz` and `foo` exists but has no folder child named `bar`, this error will complain about `foo/bar` instead of just `bar`.